src/core/qgsjsonutils.cpp

```diff
@@ -315,28 +315,69 @@ QString QgsJsonUtils::exportAttributes( const QgsFeature &feature, QgsVectorLaye
 
 QVariantList QgsJsonUtils::parseArray( const QString &json, QVariant::Type type )
 {
-  QJsonParseError error;
-  const QJsonDocument jsonDoc = QJsonDocument::fromJson( json.toUtf8(), &error );
+  QString errorMessage;
   QVariantList result;
-  if ( error.error != QJsonParseError::NoError )
+  try
   {
-    QgsLogger::warning( QStringLiteral( "Cannot parse json (%1): %2" ).arg( error.errorString(), json ) );
-    return result;
-  }
-  if ( !jsonDoc.isArray() )
-  {
-    QgsLogger::warning( QStringLiteral( "Cannot parse json (%1) as array: %2" ).arg( error.errorString(), json ) );
-    return result;
+    const auto jObj { json::parse( json.toStdString() ) };
+    if ( ! jObj.is_array() )
+    {
+      throw json::parse_error::create( 0, 0, QStringLiteral( "JSON value must be an array" ).toStdString() );
+    }
+    for ( const auto &item : jObj )
+    {
+      // Create a QVariant from the array item
+      QVariant v;
+      if ( item.is_number_integer() )
+      {
+        v = item.get<int>();
+      }
+      else if ( item.is_number_unsigned() )
+      {
+        v = item.get<unsigned>();
+      }
+      else if ( item.is_number_float() )
+      {
+        // Note: it's a double and not a float on purpose
+        v = item.get<double>();
+      }
+      else if ( item.is_string() )
+      {
+        v = QString::fromStdString( item.get<std::string>() );
+      }
+      else if ( item.is_null() )
+      {
+        // do nothing: the default is null
```

Comment:
Maybe this would be better to have NULL instead of invalid?
```
v = QVariant( QVariant::Int );
```

Reply:
@m-kuhn 
Like this?
```
  // Type not specified
  QCOMPARE( QgsJsonUtils::parseArray( R"([null, null])" ), QVariantList() << QVariant() << QVariant() );
  // Convert to int
  QCOMPARE( QgsJsonUtils::parseArray( R"([null, null])", QVariant::Type::Int ), QVariantList() << QVariant(QVariant::Type::Int) << QVariant(QVariant::Type::Int) );
```

Reply:
That's what I was thinking:

```
for ( const QVariant &value : values )
{
  QVERIFY( value.isNull() );
  QVERIFY( value.isValid() );
}
```

Reply:
@m-kuhn that won't work for `QgsJsonUtils::parseArray( R"([null, null])" )` because we if we do not know the expected type we cannot set it and `QVERIFY( QVariant().isValid() );` will fail.

Reply:
... but it's fine in the other case though.

Reply:
Yes, that's also the case when e.g. converting `qgis.core.NULL` to a `QVariant`, we just use `QVariant::Int` as fallback type (meaning possibly wrong type information under these circumstances).

Reply:
@m-kuhn please have a look to my last commit, if that interprets your thoughts.

Reply:
Yep, looking great. Just left a question that popped up, but good like this :+1: